lib/route-generator/RouteFile.ts

```diff
@@ -0,0 +1,282 @@
+import { parse, normalize, join } from "node:path";
+
+import type { Tagged } from "type-fest";
+import ts, { SyntaxKind } from "typescript";
+
+export type TypeLiteralText = Tagged<string, "ts.TypeLiteralNode">;
+
+export class RouteFile {
+	constructor(public readonly filename: string, public readonly fileContents: string) {
+		this.sourceFile = ts.createSourceFile(filename, fileContents, ts.ScriptTarget.Latest);
+
+		// Extract the nodes from the source file needed for further processing
+		for (const statement of this.sourceFile.statements) {
+			if (
+				ts.isImportDeclaration(statement) &&
+				statement.importClause?.namedBindings &&
+				ts.isNamedImports(statement.importClause.namedBindings) &&
+				statement.importClause.namedBindings.elements.some((element) => {
+					const t = (element.propertyName ?? element.name).text;
+					return t === "GetDataArgs" || t === "Props" || t === "PropsWithChildren";
+				})
+			) {
+				this.importFromGeneratedModuleNodeCandidates.push(statement);
+			} else if (ts.isFunctionDeclaration(statement)) {
+				if (statement.name && statement.name.getText(this.sourceFile) === "getData") {
+					this.getDataFunctionNode = statement;
+					this.declarations.getData = true;
+				} else if (
+					statement.name &&
+					statement.name.getText(this.sourceFile) === "LoadingState" &&
+					hasExportModifier(statement)
+				) {
+					this.declarations.LoadingState = true;
+				} else if (
+					statement.modifiers?.at(0)?.kind === SyntaxKind.ExportKeyword &&
+					statement.modifiers?.at(1)?.kind === SyntaxKind.DefaultKeyword
+				) {
+					this.declarations.default = true;
+				}
+			} else if (ts.isVariableStatement(statement) && hasExportModifier(statement)) {
+				if (
+					statement.declarationList.declarations.some(
+						(declaration) => declaration.name.getText(this.sourceFile) === "getData"
+					)
+				) {
+					this.getDataFunctionNode = statement;
+					this.declarations.getData = true;
+				} else if (
+					statement.declarationList.declarations.some(
+						(declaration) => declaration.name.getText(this.sourceFile) === "redirect"
+					)
+				) {
+					this.declarations.redirect = true;
+				} else if (
+					statement.declarationList.declarations.some(
+						(declaration) => declaration.name.getText(this.sourceFile) === "LoadingState"
+					)
+				) {
+					this.declarations.LoadingState = true;
+				}
+			} else if (
+				ts.isTypeAliasDeclaration(statement) &&
+				statement.name.text === "QueryParams" &&
+				ts.isTypeLiteralNode(statement.type) &&
+				hasExportModifier(statement)
+			) {
+				this.declarations.QueryParams = true;
+				this.queryParamsNode = statement.type;
+			}
+		}
```

Comment:
Without knowing the exact terminology of the TS AST, one can only guess which constructs are being searched for here. A description/example would probably be useful here.

Maybe introduce a helper for this piece of code:
```ts
					statement.declarationList.declarations.some(
						(declaration) => declaration.name.getText(this.sourceFile) === "xyz"
					)
```
While not that important, this would increase readability.

Reply:
Added an example for each statement